dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ExceptionDeserializingErrorDecoder.java

```diff
@@ -0,0 +1,293 @@
+                    } catch (MalformedURLException e) {
+                        log.error(
+                                "Failed to parse location header for QosException.RetryOther",
+                                UnsafeArg.of("locationHeader", locationHeader),
+                                e);
+                    }
+                } else {
+                    log.error("Retrieved HTTP status code 308 without Location header, cannot perform "
+                            + "redirect. This appears to be a server-side protocol violation.");
+                }
+            }
+            case 429 -> {
+                return Optional.of(response.getFirstHeader(HttpHeaders.RETRY_AFTER)
+                        .map(Longs::tryParse)
+                        .map(Duration::ofSeconds)
+                        .map(duration -> QosException.throttle(qosReason(response), duration))
+                        .orElseGet(() -> QosException.throttle(qosReason(response))));
+            }
+            case 503 -> {
+                return Optional.of(QosException.unavailable(qosReason(response)));
+            }
+        }
+        return Optional.empty();
+    }
+
+    private RuntimeException decodeInternal(Response response) {
+        // Check if the response code indicates an error, and if so, handle it accordingly.
+        Optional<RuntimeException> maybeException = checkCode(response);
+        if (maybeException.isPresent()) {
+            return maybeException.get();
+        }
+        int code = response.code();
+        byte[] body;
+        try (InputStream bodyStream = response.body()) {
+            body = bodyStream.readAllBytes();
+        } catch (NullPointerException | IOException e) {
+            UnknownRemoteException exception = new UnknownRemoteException(code, "<unparseable>");
+            exception.initCause(e);
+            return exception;
+        }
+
+        Optional<String> contentType = response.getFirstHeader(HttpHeaders.CONTENT_TYPE);
+        if (contentType.isEmpty() || !Encodings.matchesContentType(JSON_ENCODING.getContentType(), contentType.get())) {
+            return new UnknownRemoteException(code, new String(body, StandardCharsets.UTF_8));
+        }
+
+        try {
+            String errorName = extractErrorName(body);
+            if (errorName == null) {
+                return createRemoteException(body, code);
+            }
+            DeserializerExceptionPair<?> deserializerExceptionPair = errorNameToExceptionDeserializerMap.get(errorName);
+            if (deserializerExceptionPair == null) {
+                return createRemoteException(body, code);
+            }
+            // Attempt to deserialize the error using the deserializer for the specific exception type.
+            try {
+                AbstractSerializableError<?> error =
+                        deserializerExceptionPair.deserializer().deserialize(new ByteArrayInputStream(body));
```

Comment:
I'm not sure what the generated code will look like, but assuming the exception constructor won't take `AbstractSerializableError`, but a specific subclass, shouldn't we cast this to the exact error type? Or is that actually going to be done at runtime regardless? (if so, this is probably worth a comment)